core/eval.go

```diff
@@ -135,53 +135,121 @@ func evalGET(args []string) []byte {
 	}
 
 	key := args[0]
+	// Default path is root if not specified
+	path := "$"
+	if len(args) > 1 {
+		path = args[1]
+	}
+
+	// Retrieve the object from the database
 	obj := Get(key)
-	// Return nil if the key doesn't exist
-	if obj == nil {
+	if obj == nil || hasExpired(obj) {
 		return RESP_NIL
 	}
 
+	// Check if the object is of JSON type
 	objType, _ := ExtractTypeEncoding(obj)
 	if objType != OBJ_TYPE_JSON {
 		return Encode(errors.New("WRONGTYPE Operation against a key holding the wrong kind of value"), false)
 	}
 
-	jsonValue, ok := obj.Value.(*fastjson.Value)
-	if !ok {
-		return Encode(errors.New("ERR internal error: stored value is not a valid JSON"), false)
+	jsonData := obj.Value
+
+	// If path is root, return the entire JSON
+	if path == "$" {
+		resultBytes, err := json.Marshal(jsonData)
+		if err != nil {
+			return Encode(errors.New("ERR could not serialize result"), false)
+		}
+		return Encode(string(resultBytes), false)
 	}
 
-	return Encode(string(jsonValue.MarshalTo(nil)), false)
+	// Parse the JSONPath expression
+	expr, err := jp.ParseString(path)
+	if err != nil {
+		return Encode(errors.New("ERR invalid JSONPath"), false)
+	}
+
+	// Execute the JSONPath query
+	results := expr.Get(jsonData)
+	if len(results) == 0 {
+		return RESP_NIL
+	}
+
+	// Serialize the result
+	var resultBytes []byte
+	if len(results) == 1 {
+		resultBytes, err = json.Marshal(results[0])
+	} else {
+		resultBytes, err = json.Marshal(results)
+	}
+	if err != nil {
+		return Encode(errors.New("ERR could not serialize result"), false)
+	}
+	return Encode(string(resultBytes), false)
 }
 
-// evalJSONSET stores a JSON value at the specified key
-// args must contain at least the key, path (unused in this implementation), and JSON string
-// Returns encoded error response if incorrect number of arguments
-// Returns encoded error if the JSON string is invalid
-// Returns RESP_OK if the JSON value is successfully stored
 func evalJSONSET(args []string) []byte {
+	// Check if there are enough arguments
 	if len(args) < 3 {
 		return Encode(errors.New("ERR wrong number of arguments for 'JSON.SET' command"), false)
 	}
 
 	key := args[0]
-	// Note: args[1] (path) is ignored in this implementation
+	path := args[1]
 	jsonStr := args[2]
 
 	// Parse the JSON string
-	v, err := parser.Parse(jsonStr)
-	if err != nil {
-		return Encode(errors.New("ERR invalid JSON"), false)
+	var jsonValue interface{}
+	if err := json.Unmarshal([]byte(jsonStr), &jsonValue); err != nil {
+		return Encode(fmt.Errorf("ERR invalid JSON: %v", err), false)
+	}
+
+	// Retrieve existing object or create new one
+	obj := Get(key)
+	var rootData interface{}
+
+	if obj == nil {
+		// If the key doesn't exist, create a new object
+		if path != "$" {
+			rootData = make(map[string]interface{})
+		} else {
+			rootData = jsonValue
+		}
+	} else {
+		// If the key exists, check if it's a JSON object
+		objType, _ := ExtractTypeEncoding(obj)
```

Comment:
Use `assertType` and `assertEncoding` methods for type and encoding verification.